src/classlibnative/bcltype/grisu3.cpp

```diff
@@ -0,0 +1,376 @@
+    CachedPower(mk, &cmk, &decimalExponent);
+
+    // Step 3: Scale the w with the cached power of ten.
+
+    DiyFp D;
+    DiyFp::Multiply(w, cmk, D);
+
+    // Step 4: Generate digits.
+
+    int kappa;
+    int length;
+    bool isSuccess = DigitGen(D, count, digits, &length, &kappa);
+    if (isSuccess)
+    {
+        digits[count] = 0;
+        *dec = length - decimalExponent + kappa;
+    }
+
+    return isSuccess;
+}
+
+bool Grisu3::RoundWeed(wchar_t* buffer,
+    int len,
+    UINT64 rest,
+    UINT64 tenKappa,
+    UINT64 ulp,
+    int* kappa)
+{
+    _ASSERTE(rest < tenKappa);
+
+    // 1. tenKappa <= ulp: we don't have an idea which way to round.
+    // 2. Even if tenKappa > ulp, but if tenKappa <= 2 * ulp we cannot find the way to round.
+    // Note: to prevent overflow, we need to use tenKappa - ulp <= ulp.
+    if (tenKappa <= ulp || tenKappa - ulp <=  ulp)
+    {
+        return false;
+    }
+
+    // tenKappa >= 2 * (rest + ulp). We should round down.
+    // Note: to prevent overflow, we need to check if tenKappa > 2 * rest as a prerequisite.
+    if ((tenKappa - rest > rest) && (tenKappa - 2 * rest >= 2 * ulp))
+    {
+        return true;
+    }
+
+    // tenKappa <= 2 * (rest - ulp). We should round up.
+    // Note: to prevent overflow, we need to check if rest > ulp as a prerequisite.
+    if ((rest > ulp) && (tenKappa <= (rest - ulp) || (tenKappa - (rest - ulp) <= (rest - ulp))))
+    {
+        // Find all 9s from end to start.
+        buffer[len - 1]++;
+        for (int i = len - 1; i > 0; --i)
+        {
+            if (buffer[i] != L'0' + 10)
+            {
+                // We end up a number less than 9.
+                break;
+            }
+
+            // Current number becomes 0 and add the promotion to the next number.
+            buffer[i] = L'0';
+            buffer[i - 1]++;
+        }
+
+        if (buffer[0] == L'0' + 10)
+        {
+            // First number is '0' + 10 means all numbers are 9.
+            // We simply make the first number to 1 and increase the kappa.
+            buffer[0] = L'1';
+            (*kappa) += 1;
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+bool Grisu3::DigitGen(const DiyFp& mp, int count, wchar_t* buffer, int* len, int* K)
+{
+    // Split the input mp to two parts. Part 1 is integral. Part 2 can be used to calculate
+    // fractional.
+    //
+    // mp: the input DiyFp scaled by cached power.
+    // K: final kappa.
+    // p1: part 1.
+    // p2: part 2.
+
+    _ASSERTE(count > 0);
+    _ASSERTE(buffer != NULL);
+    _ASSERTE(len != NULL);
+    _ASSERTE(K != NULL);
+    _ASSERTE(mp.e() >= ALPHA && mp.e() <= GAMA);
+
+    UINT64 ulp = 1;
+    DiyFp one = DiyFp(static_cast<UINT64>(1) << -mp.e(), mp.e());
+    UINT32 p1 = static_cast<UINT32>(mp.f() >> -one.e());
+    UINT64 p2 = mp.f() & (one.f() - 1);
+
+    if (p2 == 0 && (count >= 11 || p1 < m_cachedPowerOfTen[count - 1]))
+    {
+        return false;
+    }
```

Comment:
could you please add a detailed comment on this part telling our error prediction for the algorithm?

Reply:
Sure I'll add the comments.